File: 40-smallest-number-after-removing-k-digits/JavaScript/solution1.js
// Time complexity: O(n)
// Space complexity: O(n)

function smallestAfterRemoving(num, k) {
    if (k == num.length) return '0';
    let stack = [];
    for (let digit of num) {
        while (stack.length > 0 && digit < stack[stack.length - 1] && k > 0) {
            stack.pop();
            k--;
        }
        stack.push(digit);
    }

    // there still have number and it's because the number is sorted
    // like 12345
    while (k-- > 0) stack.pop();
	
    // to remove leading zeros like 000235
    stack = stack.reverse();
    while (stack.length > 0 && stack[stack.length - 1] == '0') stack.pop();
    stack = stack.reverse();

    return stack.length > 0 ? stack.join('') : '0';
}

File: 40-smallest-number-after-removing-k-digits/Python/solution1.py
# Time complexity: O(n)
# Space complexity: O(n)

def smallestAfterRemoving(num, k):
  if k == len(num):
    return "0"
  stack = []
  for digit in num:
    while len(stack) > 0 and digit < stack[-1] and k > 0:
      stack.pop()
      k -= 1
    stack.append(digit)    
  while k > 0:
    stack.pop()
    k -= 1
  stack = stack[::-1]
  while len(stack) > 0 and stack[-1] == "0":
    stack.pop()
  stack = stack[::-1]
  return "".join(stack) if len(stack) > 0 else "0"


